Fixed #35784 -- Extended HttpResponseRedirectBase to preserve the original HTTP method with status codes 307 and 308. (#18616)
django/django: docs/topics/http/shortcuts.txt

   * An absolute or relative URL, which will be used as-is for the redirect
     location.

   By default, this issues a temporary redirect (status code 302).

   If ``permanent=True``, this issues a permanent redirect (status code 301).

>    If ``preserve_request_method=True``, the method and body of the original
>    request are reused to perform the redirected request. Temporary redirects
>    then have a 307 status code and permanent redirects have a 308 status code.

Reviewer: This reads a bit odd in that Django does not do anything, the one that uses a status code to issue a new request to the redirected Location is the caller (user agent). This just tells in the response "you should go to this other place and tell them the same things you told me".

Proposal:
```suggestion
   If ``preserve_request_method=True``, the redirect instructs the user agent
   to reuse the method and body of the original request when performing the
   redirected request to the ``Location`` value. Temporary redirects then use a
    307 status code, and permanent redirects use a 308 status code.
```

   This is illustrated in the following table:

>    ================  =========  =======================
>    HTTP status code  permanent  preserve_request_method
>    ================  =========  =======================
>    301               ``True``   ``False``
>    302               ``False``  ``False``
>    307               ``False``  ``True``
>    308               ``True``   ``True``
>    ================  =========  =======================

Reviewer: @sarahboyce @lorinkoz After a third iteration on these docs, I'm proposing to move the HTTP status code column to the end, so the left to right reading of this flows just like the function description does:

"when params have these values, the resulting HTTP status code is 30X"

```suggestion
   =========  ================ ================
   permanent  preserve_request HTTP status code
   =========  ================ ================
   ``True``   ``False``        301
   ``False``  ``False``        302
   ``False``  ``True``         307
   ``True``   ``True``         308
   =========  ================ ================
```

>    By default issues a temporary redirect; pass ``permanent=True`` to issue a
>    permanent redirect.
> 
>    If ``preserve_method=True``, the redirect preserves the original HTTP method.

Reviewer 2: I think we could make this clearer with a table
```suggestion
   By default, this issues a temporary redirect (status code 302).

   If ``permanent=True``, this issues a permanent redirect (status code 301).

   If ``preserve_method=True``, the method and body of the original request are
   reused to perform the redirected request. Temporary redirects then have a
   307 status code and permanent redirects have a 308 status code.

   This is illustrated in the following table:

   ================  =========  ===============
   HTTP status code  permanent  preserve_method
   ================  =========  ===============
   301               ``True``   ``False``
   302               ``False``  ``False``
   307               ``False``  ``True``
   308               ``True``   ``True``
   ================  =========  ===============
```

   By default issues a temporary redirect; pass ``permanent=True`` to issue a
   permanent redirect.

   By passing ``preserve_method=True`` the response signals the user agent to
   preserve the HTTP method when doing the redirection.

> 

Author: How do I mark this argument as added in version 5.2?

Reviewer 3: Do you mean `.. versionchanged:: 5.2`? Is this an example of what you're looking for?https://github.com/django/django/blob/6765b6adf924c1bc8792a4a454d5a788c1abc98e/docs/howto/error-reporting.txt#L287

Author: Yes, but how to use it for an added argument and not the whole function?

Reviewer 3: I'm not too sure. I did a search for `versionchanged`, `versionadded`, `argument.*added`, but I haven't seen an example yet that adds it to the argument. There are several examples in [testing tools](https://docs.djangoproject.com/en/5.1/topics/testing/tools/#django.test.Client.head) and elsewhere that apply the version at the function level. Would it be fine as you have it at the function level? Or would you want to add a `.. note` for the argument?

Reviewer: I think it will suffice if the `.. versionchanged:: 5.2` is moved to the bottom of this section.

@lorinkoz it would be ideal if you could add an example in the following section.

   By default issues a temporary redirect; pass ``permanent=True`` to issue a
   permanent redirect.

>    By passing ``preserve_method=True`` the response signals the user agent to
>    preserve the HTTP method when doing the redirection.

Reviewer: How about:
```suggestion
   If ``preserve_method=True``, the redirect preserves the original HTTP method.
```

        obj = MyModel.objects.get(...)
        return redirect(obj, permanent=True)

Additionally, the ``preserve_method`` argument can be used to preserve the
original HTTP method::

    def my_view(request):
        ...
        obj = MyModel.objects.get(...)

>         return redirect(obj, preserve_method=True)

Author: Standing by on this one.

        obj = MyModel.objects.get(...)
        return redirect(obj, permanent=True)

Additionally, the ``preserve_request_method`` argument can be used to preserve
the original HTTP method::

    def my_view(request):
        ...
        obj = MyModel.objects.get(...)

>         return redirect(obj, preserve_request_method=True)

Reviewer: I think this example should be grown to use a POST request, which is really where the 307/308 status shine. I'll add something to this effect.
